### ISSUE TITLE
"DNS" equivalent for usernames

### ISSUE THREAD

**Issue by rozeboosje** [mastodon/mastodon#24266]
### Pitch

Having lost my first Mastodon account when the Server on which it was hosted went kaputt I think there may be an argument to be made for some kind of Server Independent user handle. Akin to how DNS resolves your.website.whatever to an IP address, a Server Independent handle, for example in the form @@myusername could be assigned to a person when they sign up, and then if @username1@server.one dies for whatever reason, all they need to do is create a new account @username2@server.two and link it to @@username

I am aware that you can migrate accounts but that only works when your old server still works. In my case the old server just died unexpectedly one day, and it never came back.

Of course this would mean you'd lose your posts, but at least it may be easier for your followers to keep following you that way? Perhaps even some converse functionality to allow you recover the list of people you followed?

### Motivation

Disaster recovery. People whose server goes byebye without warning.

**Comment by DBailey635** [comment]
Would this introduce a single point of failure / hacker target if centralised? Any solution should be decentralised, maybe using blockchain to ensure the master record cannot be easily tampered with.

**Comment by rozeboosje** [comment]
Yeah.... while I am actually a developer myself I wouldn't consider myself knowledgeable enough in this specific technology to be able to come up with explicit suggestions on how to implement it

**Comment by trwnh** [comment]
you could do this with DNS itself, if you wanted to? like a TXT record with your handle or id. but you could also do this with Webfinger. you just need a Webfinger server that is trusted to be "authoritative". any webfinger servers caching the results also need to be trusted to not lie, or otherwise you'd be open to cache poisoning attacks. and you'd also need to know when to invalidate cache because of some update at the source that needs to propagate (an infamously hard problem in computer science).

currently, the only webfinger server you can trust is the one on the same domain/origin/FQDN, since it is assumed to be authoritative according to a same-origin policy -- the webfinger endpoint for something.trwnh.com should be trusted for resources on something.trwnh.com, basically. it is assumed to be always up-to-date and correct. once you introduce propagation, though, you introduce additional trust issues related to the propagation and caching. it's like asking me directly about the name of my upcoming book, versus asking my neighbor. my neighbor might not know, or might have it wrong, or might be lying, or might have outdated info.

so in practical terms, say we trusted `accounts.joinmastodon.org` to say that `acct:a@trwnh.com` is currently being hosted at `mastodon.social/users/trwnh`. someone starting at `trwnh.com` needs the lookup for `acct:a@trwnh.com` to be forwarded to `accounts.joinmastodon.org`, and any changes at `trwnh.com` need to notify `accounts.joinmastodon.org` as well. we also need to know when i've moved away from `mastodon.social/users/trwnh`.

alternatively, we can establish a "mastodon-wide" username by having `accounts.joinmastodon.org` store centralized identity mappings. you would still need to know when the user has moved to a different server, but you eliminate the other domain name from the equation at the cost of centralizing on the joinmastodon namespace and authority. see also: bluesky/atp's "placeholder server" for looking up did:plc: identifiers; Matrix's "third party identifier" (3PID) system which lets you look up emails and phone numbers and so on from a user directory. we could have several such user directories, but the problem is deciding which one is authoritative and under which namespace.
